crates/ra_hir_ty/src/expr.rs

```diff
@@ -310,3 +315,112 @@ pub fn record_pattern_missing_fields(
     }
     Some((variant_def, missed_fields, exhaustive))
 }
+
+pub struct UnsafeExpr {
+    expr: ExprId,
+    inside_unsafe_block: bool,
+}
+
+impl UnsafeExpr {
+    fn new(expr: ExprId) -> Self {
+        Self { expr, inside_unsafe_block: false }
+    }
+}
+
+pub fn unsafe_expressions(
+    db: &dyn HirDatabase,
+    infer: &InferenceResult,
+    def: DefWithBodyId,
+) -> Vec<UnsafeExpr> {
+    let mut unsafe_exprs = vec![];
+    let mut unsafe_block_scopes = vec![];
+    let body = db.body(def);
+    let expr_scopes = db.expr_scopes(def);
+    for (id, expr) in body.exprs.iter() {
+        match expr {
+            Expr::Unsafe { body } => {
+                if let Some(scope) = expr_scopes.scope_for(*body) {
```

Comment:
I don't know about using scopes for this. I don't think every block expression necessarily creates a new scope? As far as I can see, we only create new scopes when new names are introduced.

As an alternative solution that also doesn't involve walking, I could imagine creating a map from exprs to their parents in `Body` during lowering. You'd really just need to walk up and check whether there's an unsafe block somewhere in the expression's ancestors.

Reply:
I'll follow up on this and the rest after work today.

Reply:
Done

Reply:
> should expressions have parents? I don't really have a great answer here, but my gut feeling is that an extra walk is probably better than having parent links, in this case

Hmm, expressions have an identity, to which e.g. their inferred type is tied (so it would be impossible for the same expression to have multiple parents) -- so they have parents either way, the question is just whether there's an efficient way to get them, in my opinion.

Reply:
That's a very good point about identity, yeah. I think we should add parents if that's convenient then!